pkg/common/telemetry/m3.go

```diff
@@ -0,0 +1,208 @@
+	if err != nil {
+		return nil, err
+	}
+
+	scopeOpts := tally.ScopeOptions{
+		CachedReporter: r,
+	}
+
+	reportEvery := time.Second
+	scope, closer := tally.NewRootScope(scopeOpts, reportEvery)
+	sink := &m3Sink{
+		closer: closer,
+		scope: scope,
+		timerGranularity: timerGranularity,
+		enableTypePrefix: enableTypePrefix,
+	}
+
+	return sink, nil
+}
+
+func newM3TestSink(scope tally.Scope, enableTypePrefix bool) *m3Sink {
+	return &m3Sink{
+		scope: scope,
+		enableTypePrefix: enableTypePrefix,
+	}
+}
+
+func (m *m3Sink) SetGauge(key []string, val float32) {
+	m.setGauge(key, val, m.scope)
+}
+
+func (m *m3Sink) SetGaugeWithLabels(key []string, val float32, labels []Label) {
+	subscope := m.subscopeWithLabels(labels)
+	m.setGauge(key, val, subscope)
+}
+
+// Not implemented for m3
+func (m *m3Sink) EmitKey(key []string, val float32) {}
+
+// Counters should accumulate values
+func (m *m3Sink) IncrCounter(key []string, val float32) {
+	m.incrCounter(key, val, m.scope)
+}
+
+func (m *m3Sink) IncrCounterWithLabels(key []string, val float32, labels []Label) {
+	subscope := m.subscopeWithLabels(labels)
+	m.incrCounter(key, val, subscope)
+}
+
+// Samples are for timing information, where quantiles are used
+func (m *m3Sink) AddSample(key []string, val float32) {
+	m.addSample(key, val, m.scope)
+}
+
+func (m *m3Sink) AddSampleWithLabels(key []string, val float32, labels []Label) {
+	subscope := m.subscopeWithLabels(labels)
+	m.addSample(key, val, subscope)
+}
+
+func (m *m3Sink) subscopeWithLabels(labels []Label) tally.Scope {
+	tags := labelsToTags(labels)
+	return m.scope.Tagged(tags)
+}
+
+// Flattens the key for formatting, removes spaces
+func (m *m3Sink) flattenKey(parts []string) string {
+	var keyParts []string
+	if m.enableTypePrefix {
+		// Ignore service name and type of metric as part of metric name,
+		// i.e. prefer "foo_bar" to "service_counter_foo_bar"
+		keyParts = parts[2:]
+	} else {
+		keyParts = parts[1:]
+	}
+
+	return strings.Join(keyParts, "_")
+}
+
+func labelsToTags(labels []Label) map[string]string {
+	tags := make(map[string]string, len(labels))
+	for _, l := range labels {
+		tags[l.Name] = l.Value
+	}
+
+	return tags
+}
+
+func (m *m3Sink) setGauge(key []string, val float32, scope tally.Scope) {
+	gauge := m.getGauge(key, scope)
+	val64 := float64(val)
+	gauge.Update(val64)
+}
+
+func (m *m3Sink) getGauge(key []string, scope tally.Scope) tally.Gauge {
+	flattenedKey := m.flattenKey(key)
+	return scope.Gauge(flattenedKey)
+}
+
+func (m *m3Sink) incrCounter(key []string, val float32, scope tally.Scope) {
+	counter := m.getCounter(key, scope)
+	val64 := int64(val)
+	counter.Inc(val64)
+}
+
+func (m *m3Sink) getCounter(key []string, scope tally.Scope) tally.Counter {
+	flattenedKey := m.flattenKey(key)
+	return scope.Counter(flattenedKey)
+}
+
+func (m *m3Sink) addSample(key []string, val float32, scope tally.Scope) {
+	flattenedKey := m.flattenKey(key)
+	if m.enableTypePrefix && key[1] == "timer" {
```

Comment:
is `key` guaranteed to have at least two elements?

Reply:
there will always at least be `spire_agent/server`, possible metric prefix, and the at least 1 key (though usually more) provided at the time telemetry was called
just for sanity might be good to still have a `len` check anyway :) maybe logging too when `len` is unexpectedly short?

Reply:
Save any breaking changes from `armon/go-metrics` with respect to its handling of the `ServiceName` and `EnableTypePrefix` configuration properties, `key` is guaranteed to have at least two elements as @amoore877 mentioned. The `ServiceName`, e.g. `spire_agent` or `spire_server`, will be the first element and the metric type, e.g. `counter`, `sample`, will always be the second element.